Bonuses - API calls/routes/DB models

KimarieHB opened the pull request.

Commit:
Establish one successful API call with sample data

Commit:
Update get request for bundle bonus to take req params

Commit:
Close mongoose/db connection after seeding

Commit:
Update successful seed statement with entries quantity

Commit:
Separate DB connection and queries from server

into database/index.js

Commit:
Add API call to get a single bonus item from db

Commit:
Add API call for specific track to be played

returns entire album, update some calls to use req.query

Commit:
Successfully pass song title to db query and retireve songs data

investigating project fields to return one song from db instead of complete album object

Commit:
Tidy after linting db and server files

Commit:
Merge branch 'master' of github.com:rpt23-team-feedback/kimarie-bonuses into apiCalls

Commit:
Complete all API calls to db

Commit:
Tidy server/index.js and db/index.js to be ready fo PR

Commit:
Change song to song_url in schema

gektin commented on database/index.js:
```diff
@@ -0,0 +1,65 @@
+const mongoose = require('mongoose');
+const Soundtrack = require('./models/soundtracks.js');
```
I think this would read more clearly if your variable name was also pluralized, but I don't feel too strongly.

gektin commented on database/index.js:
```diff
@@ -0,0 +1,65 @@
+  dbName: 'bonuses',
+  useNewUrlParser: true,
+  useUnifiedTopology: true
+}).then( () => {
+  console.log('Connection to MongoDB successful!');
+}).catch( (err) => {
```
I don't believe it's best practice to add a padding space after an open parenthesis.

KimarieHB replied:
Space removed.

gektin commented on database/index.js:
```diff
@@ -0,0 +1,65 @@
+});
+
+// Get bonus associated with bundle
+const getBonuses = (bundleId, callback) => {
```
It's a bit strange to see you using promises at the top of this file in connecting to your db and then switching here to a callback style. Are you intending to use callbacks throughout your code, or are you planning on using promises? Either way is fine, I believe, but you should pick one and stick with it.

KimarieHB replied:
If there is time after everything is running I will switch my db functions to the wonderful world of promises. For now, I am feeling the necessity of getting some basic, clean working code in my repo. Moving forward I will be more consistent with this.

KimarieHB replied:
Everything, meaning my entire app.

gektin commented on database/index.js:
```diff
@@ -0,0 +1,65 @@
+
+// Get individual bonus item info
+const getBonusItem = (title, callback) => {
+  Soundtrack.find({ "bonus_info.title": `${title}` }, (err, data) => {
```
I read this as invoking this `find` function on a key called `bonus_info.title`, does mongoose expect you to enter in key names in double quotes? I believe we're supposed to be using single quotes where possible. I also am not sure that you need the string literal around the value (see your use of `bundleId` without the string literal on line 17).

KimarieHB replied:
Leftovers from earlier problems. Switched to single quotes and removed string literal.

gektin commented on database/index.js:
```diff
@@ -0,0 +1,65 @@
+          song = {
+            trackNo: track.track_number,
+            name: track.name,
+            songUrl: track.song
```
I believe we are supposed to use trailing commas after the last item in a multi-line object/array.

KimarieHB replied:
Converted. Dangled commas all over the place!

gektin commented on server/index.js:
```diff
@@ -3,6 +3,7 @@ const app = express();
 const parser = require('body-parser');
 const db = require('../database/index.js');
+
 let port = 3031;
```
Do you change your port anywhere? Is it best practice to use a `let` for our ports? This seems like it could be a `const` to me.

KimarieHB replied:
Upon deployment port could change.

gektin commented on server/index.js:
```diff
@@ -3,6 +3,7 @@ const app = express();
+
 let port = 3031;
 
 if (port === null || port === '') {
```
I'm confused by this code. Why is this here if you're setting your port above?

KimarieHB replied:
Changed.

gektin commented on server/index.js:
```diff
@@ -22,3 +23,42 @@ app.use(parser.json());
+app.get('/bonus-title', (req, res) => {
+  let selectedTitle = req.query.title;
+
+  db.getBonusItem(selectTitle, (err, results) => {
```
I think this should be `selectedTitle` as the first passed in argument here based on the declaration on line 42.

KimarieHB replied:
Changed it to selectedTitle.

gektin commented on server/index.js:
```diff
@@ -22,3 +23,42 @@ app.use(parser.json());
+
+// To render individual bonus upon click selection
+app.get('/bonus-title', (req, res) => {
+  let selectedTitle = req.query.title;
```
Is there a reason to use query params in one case, here on line 42, and req params above, see line 29? If someone else has to hit these routes, then it may be confusing to have one served by one kind of parameter and the other with a different kind.

KimarieHB replied:
Request params do not accommodate multi-word options (at least not the way I was testing it).

gektin replied:
camelCase?

gektin reviewed (commented):
I think your code makes sense for the most part. Most of my comments are small style notes or pointing out places where consistency falters. Of those, there's a mix between promises and callbacks that may cause more confusion for you in trying to work on your code, as well as inconsistent use of req params v. query params, which may prove confusing to other people hitting your routes.

I believe you also have a separate PR outstanding with tests for this code, but I just wanted to flag that if you don't, that you'll need to add tests.

Commit:
Merge branch 'apiCalls' of github.com:KimarieHB/kimarie-bonuses into apiCalls

KimarieHB reviewed (commented):
Updated code and will update PR

Commit:
Update code to reflect change recommendations

from PR #3 code review comments; change song to song_url at seed.js and through out all files where used.
